src/devices/ovrheadset/OVRHeadset.h

```diff
@@ -132,58 +144,74 @@ class OVRHeadset : public yarp::dev::DeviceDriver,
-    yarp::os::BufferedPort<yarp::os::Bottle>* positionPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* angularVelocityPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* linearVelocityPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* angularAccelerationPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* linearAccelerationPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* predictedOrientationPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* predictedPositionPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* predictedAngularVelocityPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* predictedLinearVelocityPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* predictedAngularAccelerationPort;
-    yarp::os::BufferedPort<yarp::os::Bottle>* predictedLinearAccelerationPort;
-    FlexImagePort*                            gui_ports;
-    
+#ifndef YARP_EXPERIMENTAL
+    yarp::os::BufferedPort<yarp::os::Bottle>* orientationPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* positionPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* angularVelocityPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* linearVelocityPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* angularAccelerationPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* linearAccelerationPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* predictedOrientationPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* predictedPositionPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* predictedAngularVelocityPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* predictedLinearVelocityPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* predictedAngularAccelerationPort{ nullptr };
+    yarp::os::BufferedPort<yarp::os::Bottle>* predictedLinearAccelerationPort{ nullptr };
+#endif
+
+    FlexImagePort* gui_ports{ nullptr };
     std::vector<guiParam> huds;
-    InputCallback* displayPorts[2];
+    InputCallback* displayPorts[2]{ nullptr, nullptr };
     ovrEyeRenderDesc EyeRenderDesc[2];
-    TextureStatic* textureLogo;
+    TextureStatic* textureLogo{ nullptr };
     ovrLayerQuad logoLayer;
-    TextureStatic* textureCrosshairs;
+    TextureStatic* textureCrosshairs{ nullptr };
     ovrLayerQuad crosshairsLayer;
-    TextureBattery* textureBattery;
+    TextureBattery* textureBattery{ nullptr };
     ovrLayerQuad batteryLayer;
-    ovrMirrorTexture mirrorTexture;
-    GLuint mirrorFBO;
+    ovrMirrorTexture mirrorTexture{ nullptr };
+    GLuint mirrorFBO{ 0 };
     ovrSession session;
     ovrHmdDesc hmdDesc;
-    GLFWwindow* window;
+    GLFWwindow* window{ nullptr };
     ovrTrackingState ts;
     ovrPoseStatef headpose;
+    ovrPoseStatef predicted_headpose;
     unsigned int guiCount;
-    bool         enableGui;
-
+    bool         enableGui{ true };
+#ifdef YARP_EXPERIMENTAL
+    std::unique_ptr<std::mutex> frameMutex{ new std::mutex };
```

Comment:
Why not simply `std::mutex` instead of `std::unique_ptr<std::mutex>` ?

Reply:
because the mutex isn't alway involved. is involved only in the case where the user don't activate the callback (and so there are two threads). otherwise only the device thread is involved. in this case there's no need for the mutex

Reply:
What if the pointer to the device is passed to another device by the yarprobotinterface using the `IMultipleWrapper` method? In that case, any `IFrameSource` method of the class could be called by any arbitrary thread.

Reply:
i see your point. i'll think a solution for it.

Reply:
ok, a solution could be to enforce the usage of the callback if a user activate it and force in this case a single thread usage of the device (which will call multiple callback, one per user).

Reply:
it's easier.. i was thinking a way to make the lock and unlock not mandatory.. if i can't find any way to do it i will always use the mutex